core/job/service/job_service.go

```diff
@@ -353,22 +363,81 @@ func (j *JobService) Validate(ctx context.Context, jobTenant tenant.Tenant, jobS
 
 	// NOTE: only check cyclic deps across internal upstreams (sources), need further discussion to check cyclic deps for external upstream
 	// assumption, all job specs from input are also the job within same project
+	jobsToValidateMap := getAllJobsToValidateMap(incomingJobs, existingJobs, toDelete)
+	identifierToJobsMap := getIdentifierToJobsMap(jobsToValidateMap)
+	for _, jobEntity := range jobsToValidateMap {
+		if _, err := j.validateCyclic(jobEntity.Job().Spec().Name(), jobsToValidateMap, identifierToJobsMap); err != nil {
+			me.Append(err)
+			break
+		}
+	}
 
-	// populate all jobs in project
-	jobsInProjectWithUpstreams, err := j.getJobsInProjectWithUpstreams(ctx, jobTenant.ProjectName())
-	if err != nil {
-		return err
+	if len(me.Errors) > 0 {
+		return me
 	}
 
-	jobMap := make(map[job.Name]*job.WithUpstream)
-	identifierToJobsMap := make(map[string][]*job.WithUpstream)
-	for _, jobEntity := range append(jobsWithUnresolvedUpstreams, jobsInProjectWithUpstreams...) {
-		jobSpecName := jobEntity.Job().Spec().Name()
-		if _, ok := jobMap[jobSpecName]; ok {
+	return nil
+}
+
+func (j *JobService) validateDeleteJobs(ctx context.Context, jobTenant tenant.Tenant, toDelete []*job.Spec, logWriter writer.LogWriter) error {
+	me := errors.NewMultiError("delete job specs check errors")
+	for _, jobToDelete := range toDelete {
+		downstreamList, err := j.repo.GetDownstreamByJobName(ctx, jobTenant.ProjectName(), jobToDelete.Name())
+		if err != nil {
+			logWriter.Write(writer.LogLevelError, fmt.Sprintf("[%s] pre-delete check for job %s failed: %s", jobTenant.NamespaceName().String(), jobToDelete.Name().String(), err.Error()))
+			me.Append(err)
 			continue
 		}
-		jobMap[jobSpecName] = jobEntity
 
+		if len(downstreamList) > 0 {
+			downstreamFullNames := job.DownstreamList(downstreamList).GetDownstreamFullNames()
+			errorMsg := fmt.Sprintf("deletion of job %s will fail. job is being used by %s", jobToDelete.Name().String(), downstreamFullNames.String())
+			logWriter.Write(writer.LogLevelError, fmt.Sprintf("[%s] %s", jobTenant.NamespaceName().String(), errorMsg))
+			me.Append(errors.NewError(errors.ErrFailedPrecond, job.EntityJob, errorMsg))
+			continue
+		}
+	}
+	return errors.MultiToError(me)
+}
+
+func getAllJobsToValidateMap(incomingJobs, existingJobs []*job.Job, toDelete []*job.Spec) map[job.Name]*job.WithUpstream {
+	me := errors.NewMultiError("validate specs errors")
+
+	jobsToValidateMap := make(map[job.Name]*job.WithUpstream)
+	for _, incomingJob := range incomingJobs {
```

Comment:
we can directly use incoming+unmodified here

Reply:
updated